lib/helpers/macos/applescript.js

```diff
@@ -0,0 +1,117 @@
+/// <reference path="./types.js" />
+
+'use strict';
+
+const { execFile } = require('child_process');
+
+const { KeyboardAction } = require('./KeyboardAction');
+const { KeyCode } = require('./KeyCode');
+const { Modifier } = require('./Modifier');
+const { validateCommand } = require('./validateCommand');
+
+/**
+ * Transform at-driver/webdriver code points into an applescript and run it.
+ * KeyCombination -> AppleKeyCodeCommand -> Applescript -> runScript
+ * @file
+ */
+
+const APPLESCRIPT_RETRY = 3;
+const APPLESCRIPT_TIMEOUT = 5;
+const APPLESCRIPT_TIMEOUT_ERROR = 'AppleEvent timed out';
+
+exports.KeyboardAction = KeyboardAction;
+
+exports.KeyCode = KeyCode;
+
+exports.Modifier = Modifier;
+
+const ScriptKind = (exports.ScriptKind = 'Applescript.Script');
+
+/**
+ * Create an Applescript object.
+ * @param {string} script
+ * @returns {Applescript.Script}
+ */
+const applescript = (exports.applescript = function (script) {
+  return {
+    kind: ScriptKind,
+    script,
+  };
+});
+
+/**
+ * Render a command into an Applescript.
+ * @param {Applescript.Command} command
+ * @returns {Applescript.Script}
+ * @throws
+ */
+exports.renderScript = function (command) {
+  validateCommand(command);
+  const script = `with timeout of ${APPLESCRIPT_TIMEOUT} seconds
+    tell application "System Events"${command.keyCodes
+      .map(
+        code => `
+        key code ${KeyCode[code].toString()}${
+          command.modifiers.length > 1
+            ? ` using {${command.modifiers.map(mod => `${Modifier[mod]} down`).join(', ')}}`
+            : command.modifiers.length === 1
+              ? ` using ${Modifier[command.modifiers[0]]} down`
+              : ``
+        }`,
+      )
+      .join('')}
+    end tell
+end timeout`;
+  return applescript(script);
+};
+
+/**
+ * @param {Applescript.Script} script
+ * @returns {Promise<string | void>}
+ * @throws
+ */
+const execScript = (exports.execScript = async function (script) {
+  return new Promise((resolve, reject) => {
+    const child = execFile('/usr/bin/osascript', [], {}, (e, stdout) => {
+      if (e) {
+        return reject(e);
+      }
+
+      if (!stdout) {
+        resolve();
+      } else {
+        resolve(stdout);
+      }
+    });
+
+    child.stdin.write(script.script);
```

Comment:
This demonstrates an overloaded definition of the word "script," potentially confusing discussion amongst maintainers. While the immediate remedy would be to introduce a distinct term for one of the two concepts (e.g. `script` and `script.source`), I don't see why we need both in the first place. As long as there is only one "ScriptKind", what is the value in a container data structure? In other words, couldn't a script just be a string?

Reply:
I think using a container is more maintainable. One thing we can do with the container that we can't do with a `string` is document the container. That type information can persist through different functions. Type analysis can give you references to what takes it as a parameter and what creates the type.

-----

I'll update the member to `source`. I really like that improvement. I was having a similar thought that calling it `script` was redundant.

Reply:
Works for me